### ISSUE TITLE
x/tools/go/buildutil: split TagsFlagDoc into multiple lines?

### ISSUE THREAD

**Issue by mvdan** [golang/go#23464]
### What version of Go are you using (`go version`)?

    go version devel +7e054553ad Tue Jan 16 15:11:05 2018 +0000 linux/amd64

### Does this issue reproduce with the latest release?

Yes.

### What did you do?

Add a `-tags` flag to a static analysis tool of mine:

```
func init() {
        flag.Var((*buildutil.TagsFlag)(&build.Default.BuildTags), "tags", buildutil.TagsFlagDoc)
}
```

Then run `mytool -h`.

### What did you expect to see?

The long flag usage line to be split in multiple lines somehow.

```
Usage of mytool:
  -tags build tags
        a list of build tags to consider satisfied during the build.
        For more information about build tags, see the description of
        build constraints in the documentation for the go/build package.
```

### What did you see instead?

```
Usage of mytool:
  -tags build tags
        a list of build tags to consider satisfied during the build. For more information about build tags, see the description of build constraints in the documentation for the go/build package
```

I initially thought that changing the string constant would be enough, for example by adding `"\n\t"` twice to split the one log line into three lines.

However, the `flag` package is not consistent about placing the usage line, so I am not sure if that is enough:

```
// Boolean flags of one ASCII letter are so common we
// treat them specially, putting their usage on the same line.
if len(s) <= 4 { // space, space, '-', 'x'.
        s += "\t"
} else {
        // Four spaces before the tab triggers good alignment
        // for both 4- and 8-space tab stops.
        s += "\n    \t"
}
```

So perhaps this should be a proposal to add special handling of long flag usage texts in the `flag` package, to wrap them around a certain around of columns. But I imagine that such a feature wouldn't be welcome, as the package is meant to be simple, and one has multiple workarounds available like using shorter texts, using a custom `Usage` func, or using a third-party flag package.

Still, it seems to me like this is a problem with either `buildutil` or `flag`.

/cc @alandonovan @shurcooL

**Comment by shurcooL** [comment]
> The long flag usage line to be split in multiple lines somehow.

Just checking, but doesn't your terminal perform automatic wrapping of long lines?

For me, I get something like this:

![image](https://user-images.githubusercontent.com/1924134/35064338-ace30dcc-fb97-11e7-96d6-f5d3ded18c16.png)

**Comment by mvdan** [comment]
Yes, but that's harder to read since there is no indentation. The end of that long line and the start of the next flag can be confused as one single paragraph. That's a good point that I didn't mention, though.

**Comment by shurcooL** [comment]
Splitting the help string into multiple lines is more viable in Go 1.10, given that https://tip.golang.org/doc/go1.10#flag was done:

> PrintDefaults now adds appropriate indentation after newlines in flag usage strings, so that **multi-line usage strings display nicely.**

_(Discovered via https://speakerdeck.com/campoy/the-state-of-go-1-dot-10?slide=23. /cc @campoy FYI.)_

**Comment by mvdan** [comment]
Yes - I saw Francesc's slides too, and the same thing came to mind. I forgot to follow up on this issue, though :)

Since the x repos tend to support the latest 2 Go releases, I think it would be okay to add the newlines once 1.10 is out. Those building with 1.9 and using that help string will get output that is slightly off, but I think that is okay.

The other option is waiting until 1.11 is out. I'm fine either way.

**Comment by spf13** [comment]
ping @alandonovan  for decision

**Comment by alandonovan** [comment]
If this problem really needs to be solved, its solution belongs in the flag package, not in every flag docstring, since the author of the flag docstring cannot know how long to make each line as this depends on the indentation applied by the flag package.

**Comment by shurcooL** [comment]
But the `flag` package can't reliably do wrapping with indentation either, not without knowing the width of the terminal where its output is being shown. Is that possible?

@rsc's decision in #20799 (which led to the Go 1.10 `flag` change 0828ec1ea8f1733d61db0893fe8c03abbcebc2e5, described at https://golang.org/doc/go1.10#flag) seemed to suggest that it's ok for the docstring to be split into multiple lines.

**Comment by alandonovan** [comment]
On 27 March 2018 at 23:51, Dmitri Shuralyov <notifications@github.com>
wrote:

> But the flag package can't reliably do wrapping with indentation either,
> not without knowing the width of the terminal where its output is being
> shown. Is that possible?
>
> @rsc <https://github.com/rsc>'s decision in #20799
> <https://github.com/golang/go/issues/20799> (which led to the Go 1.10 flag
> change 0828ec1
> <https://github.com/golang/go/commit/0828ec1ea8f1733d61db0893fe8c03abbcebc2e5>,
> described at https://golang.org/doc/go1.10#flag) seemed to suggest that
> it's ok for the docstring to be split into multiple lines.
>

Logf does something much simpler: it inserts a number of tabs after each
newline, so that on an infinitely wide terminal, multiline comments indent
nicely. It does not attempt to solve the problem of automatic line
wrapping, which, as you point out, depends on the width of the terminal.
(Under Bazel-like build systems, tests may be executed remotely in a
neutral environment, and their results are sent back to the user's machine,
so the width of the user's terminal, if it exists at all, cannot be an
input to testing.Logf.)

Changing flag to resemble Logf seems fine.  Making either guess the
terminal width does not.

**Comment by shurcooL** [comment]
> Logf does something much simpler: it inserts a number of tabs after each newline, so that on an infinitely wide terminal, multiline comments indent nicely. […] Changing flag to resemble Logf seems fine.

Oh, but that’s already done. That’s exactly the 1.10 flag change I mentioned above. Sorry if I’ve caused any confusion.

I think the decision Steve asked for is whether to wait until 1.11 before splitting `TagsFlagDoc` into multiple shorter strings (that flag will indent correctly as of 1.10) or do it now.
